lib/pulse_sensor.c

```diff
@@ -0,0 +1,332 @@
+/**
+ * Library for operating pulse one or multiple sensors and returning an result through a float pointer.
+ *
+ * Alex Tong
+ */
+
+#include "stm32f4xx_hal.h"
+#include "pulse_sensor.h"
+#include "base_types.h"
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <math.h>
+
+static PulseSensor pulseSensor[TIMER_COUNT] = {0};
+static U8 numSensors = 0;
+
+static void clear_buffer_and_reset_dma(U8 sensorNumber);
+static float convert_delta_time_to_frequency(float deltaTime, float timerPeriodSeconds);
+
+// Most useful debug variables
+#ifdef DEBUG_PS
+static U32 deltaList[MAX_DELTAS] = {0};
+static float TrackedFrequency = 0;
+#endif
+
+// Function for setting up timer with all details, use this function directly to include variable speed sampling (vss) data.
+int setup_pulse_sensor_vss(
+		TIM_HandleTypeDef* htim,
```

Comment:
It might be good to explain what all these values are in the init function. Should be able to just copy from your struct definition in the .h file